### PR TITLE
Add a Windows CI.

### PR DESCRIPTION
**Pull request by mgautierfr**
Compilation of kiwix-tools is broken on Windows.
But let's setup the CI to validate the PR (to be fixed)

### CODE REVIEW

**Comment by kelson42** [comment]
I though we have Windows CD using win2019 to ensure binaries run on older versions of Windows. But here CI is only win2022... which would be a pb.

What are version of windows used for CI, for CD? And why?

**Comment by mgautierfr** [comment]
win2019 vs win2022 is mainly about visual studio version.
Windows itself is the same.

I suppose we should use win2022 everywhere. But it seems not important as you have added a libzim CI using win2022 which compile with deps build on win2019...

**Comment by kelson42** [comment]
@mgautierfr Then why relying on win2019 at all?

**Comment by kelson42** [comment]
@mgautierfr Then why relying on win2019 at all? We should use only 2022 version everywhere.

**Comment by mgautierfr** [comment]
> Then why relying on win2019 at all? 

Where ?

File: src/searcher/kiwix-search.cpp [hunk 1]
```diff
@@ -28,88 +28,75 @@
...
-      for (const auto& r : searcher.suggest(search).getResults(0, 10) ) {
-        cout << r.getTitle() << endl;
+      auto results = searcher.suggest(pattern).getResults(0, 10);
+      for (auto it=results.begin(); it!=results.end(); it++) {
+        auto suggestionItem = *it;
+        cout << "Title : " << suggestionItem.getTitle() << endl;
+        cout << "Path : " << suggestionItem.getPath() << endl;
+        auto entry = it.getEntry();
+        cout << "Entry title : "  << entry.getTitle() << endl;
+        cout << endl;
```
Comment (veloman-yunkan) [inline src/searcher/kiwix-search.cpp:94]: This is a change in the output of `kiwix-search --suggestion`. If it's intentionaly, it better be performed in a separate commit, and I think that the full-text (non `--suggestion`) mode should be updated too.

BTW, when can **"Title :"** be different from **"Entry title :"**?

**Comment by mgautierfr** [comment]
Packages workflow because kiwix-tools cannot found the docopt dependencies.
However, I have added the dependency to `debian/control` and it is installed (from logs).
We are doing the same thing than in zim-tools (where it works).

The only difference with zim-tools is that docopt is a new dependency here. Maybe there is some kind of cache somewhere ?
@kelson42 @legoktm do you have an idea ?

File: src/server/kiwix-serve.cpp [hunk 2]
```diff
@@ -41,50 +41,47 @@
...
+ -v --verbose                            Print debug log to STDOUT
+ -V --version                            Print software version
+ -z --nodatealiases                      Create URL aliases for each content by removing the date
+ -c=<path> --customIndex=<path>          Add path to custom index.html for welcome page
+ -L=<limit> --ipConnectionLimit=<limit>  Max number of (concurrent) connections per IP [default: 0] (recommended: >= 6)
+ -k --skipInvalid                        Startup even when ZIM files are invalid (those will be skipped)
+
+Documentation:
+  Source code   https://github.com/kiwix/kiwix-tools
+  More info     https://wiki.kiwix.org/wiki/Kiwix-serve
```
Comment (veloman-yunkan) [inline src/server/kiwix-serve.cpp:83]: It makes sense to add the https://kiwix-tools.readthedocs.io/en/latest/kiwix-serve.html link

File: meson.build [hunk 3]
```diff
@@ -20,8 +20,9 @@ libzim_dep = dependency('libzim', version:'>=9.0.0', static:static_linkage)
 libzim_dep = dependency('libzim', version:'<10.0.0', static:static_linkage)
 kiwixlib_dep = dependency('kiwix', version:'>=13.0.0', static:static_linkage)
 kiwixlib_dep = dependency('kiwix', version:'<14.0.0', static:static_linkage)
+docopt_dep = dependency('docopt', static:static_linkage)
```
Comment (veloman-yunkan) [inline meson.build:27]: I would split this commit ("Port kiwix-tools to docopt.") into three - one per tool.

File: src/server/kiwix-serve.cpp [hunk 4]
```diff
@@ -422,7 +424,11 @@ int main(int argc, char** argv)
 
     if ( monitorLibrary ) {
       curLibraryFileTimestamp = newestFileTimestamp(libraryPaths);
+#ifndef _WIN32
       libraryMustBeReloaded += curLibraryFileTimestamp > libraryFileTimestamp;
+#else
+      libraryMustBeReloaded |= curLibraryFileTimestamp > libraryFileTimestamp;
+#endif
```
Comment (veloman-yunkan) [inline src/server/kiwix-serve.cpp:431]: Why is this change needed? `+=` should have the needed effect on a `bool` variable too. Or does the compiler issue a warning?

**Review by veloman-yunkan — COMMENTED** [review]
I mainly looked at the C++ code. Note that I didn't check that the usage strings were updated accurately.

File: src/searcher/kiwix-search.cpp (hunk 1 above)
Comment (mgautierfr) [reply src/searcher/kiwix-search.cpp:94]: This is not intentionnal, this is a old try/error I had on my working tree.

I will remove this.

Comment (mgautierfr) [reply src/searcher/kiwix-search.cpp:94]: > BTW, when can "Title :" be different from "Entry title :"?

If, at zim creation, user (scrapper) reimplement a `IndexData::getTitle()` which return something different than entry title (https://github.com/openzim/libzim/blob/main/include/zim/writer/item.h#L73)

File: src/server/kiwix-serve.cpp (hunk 4 above)
Comment (mgautierfr) [reply src/server/kiwix-serve.cpp:431]: > Or does the compiler issue a warning?

It is. And we are compiling with `Werror`.

**Comment by mgautierfr** [comment]
New version splits commit and small fixes of previous review are already squashed.

File: src/server/kiwix-serve.cpp (hunk 4 above)
Comment (veloman-yunkan) [reply src/server/kiwix-serve.cpp:432]: We better address that problem without conditional compilation. For example:

```c++
if ( !libraryMustBeReloaded ) { 
  libraryMustBeReloaded = curLibraryFileTimestamp > libraryFileTimestamp;
}
```

**Review by veloman-yunkan — APPROVED** [review]
Rebase-fixup-and-merge before @kelson42 merges without fixing-up :)

**Comment by kelson42** [comment]
@mgautierfr All the deb packages pass on `main`, so I guess it should be green here too.

**Comment by mgautierfr** [comment]
Yes, see this comment : https://github.com/kiwix/kiwix-tools/pull/695#issuecomment-2321179930

**Comment by kelson42** [comment]
@mgautierfr @veloman-yunkan For an unknow reason xmake seems to be missing https://github.com/kiwix/kiwix-tools/actions/runs/10634049570/job/29480511704#step:10:507

**Comment by tusharpm** [comment]
Try adding `cmake` to Build-Depends in `debian/control`. It seems meson is unable to locate the installed libdocopt-dev using pkg-config.

**Comment by kelson42** [comment]
So, now it passes and I'm going to merge.
